src/System.Console/tests/SyncTextReader.cs

```diff
@@ -0,0 +1,206 @@
+        "System.Object",
+        "Hello World",
+    };
+
+    private static void Test(string content, Action<StreamReader> action)
+    {
+        TextWriter savedStandardOutput = Console.Out;
+        TextReader savedStandardInput = Console.In;
+
+        try
+        {
+            using (MemoryStream memStream = new MemoryStream())
+            {
+                using (StreamWriter sw = new StreamWriter(memStream))
+                {
+                    sw.Write(content);
+                    sw.Flush();
+
+                    memStream.Seek(0, SeekOrigin.Begin);
+
+                    using (StreamReader sr = new StreamReader(memStream))
+                    {
+                        Console.SetIn(sr);
+                        action(sr);
+                    }
+                }
+            }
+        }
+        finally
+        {
+            Console.SetOut(savedStandardOutput);
+            Console.SetIn(savedStandardInput);
+        }
+    }
+
+    [Fact]
+    public void ReadToEnd()
+    {
+        var expected = string.Join(Environment.NewLine, s_testLines);
+        Test(expected, reader => {
+            // Given, When
+            var result = Console.In.ReadToEnd();
+
+            // Then
+            Assert.Equal(expected, result);
+            Assert.Equal(-1, Console.Read()); // We should be at EOF now.
+        });
+    }
+
+    [Fact]
+    public void ReadBlock()
+    {
+        var expected = new[] { 'H', 'e', 'l', 'l', 'o' };
+        Test(new string(expected), reader => {
+            // Given
+            var buffer = new char[5];
+
+            // When
+            var result = Console.In.ReadBlock(buffer, 0, 5);
+
+            // Then
+            Assert.Equal(5, result);
+            Assert.Equal(expected, buffer);
+            Assert.Equal(-1, Console.Read()); // We should be at EOF now. 
+        });
+    }
+
+    [Fact]
+    public void Read()
+    {
+        var expected = new[] { 'H', 'e', 'l', 'l', 'o' };
+        Test(new string(expected), reader => {
+            // Given
+            var buffer = new char[5];
+
+            // When
+            var result = Console.In.Read(buffer, 0, 5);
+
+            // Then
+            Assert.Equal(5, result);
+            Assert.Equal(expected, buffer);
+            Assert.Equal(-1, Console.Read()); // We should be at EOF now.
+        });
+    }
+
+    [Fact]
+    public void Peek()
+    {
+        Test("ABC", reader => {
+            // Given, When
+            var first = Console.In.Peek();
+            var second = Console.In.Read();
+
+            // Then
+            Assert.Equal('A', (char)first);
+            Assert.Equal('A', (char)second);
+        });
+    }
+
+    [Fact]
+    public void ReadToEndAsync()
+    {
+        var expected = string.Join(Environment.NewLine, s_testLines);
+        Test(expected, reader => {
+            // Given, When
+            var task = Console.In.ReadToEndAsync();
+            task.Wait(TimeSpan.FromSeconds(1));
```

Comment:
Result will block until the operation has completed.  You don't need this Wait, and in general we shy away from timeouts like this in tests anyway (though unfortunately lots of our tests still have them), as they can lead to random failures depending on machine conditions.